lib/nat/interface.go

```diff
@@ -23,5 +23,7 @@ type Device interface {
 	ID() string
 	GetLocalIPAddress() net.IP
 	AddPortMapping(ctx context.Context, protocol Protocol, internalPort, externalPort int, description string, duration time.Duration) (int, error)
+	TryAddPinhole(ctx context.Context, protocol Protocol, port int, description string, duration time.Duration) (int, error)
 	GetExternalIPAddress(ctx context.Context) (net.IP, error)
+	IsIPv6() bool
```

Comment:
This interface is now trying to be agnostic of the actual method that the "firewall traversal" is performed, but not doing that anymore. Leading to quite a few methods which mostly arent used (and it looks like `IsIPv6` is the same as `GetLocalIPAddress().To4() == nil`?). Is anything preventing us from simply doing the ipv6 magic inside the concrete implementation, thus not needing any changes in lib/nat? And if the "port-mapping" naming is too inaccurate for that, rename that to e.g. `AddPinhole` (pinhole seems to be popularized by ipv6, but also might be a more general term also applying to NAT?), `AddFirewallTraversal` (urgh), or a much better name of course.

Reply:
I thought about that too but I think `nat.Service` should work differently for IPv4 and IPv6. For example, the port mapping logic: With IPv4 the public IP is shared so there is a probability that the port we want is taken by some other software. Therefore, there's logic to try random ports until we find a free one. With IPv6, we have our own GUA so we just want to use our port since there's no potential for conflict. I don't really see how a common interface for these two would look like (accept external port as an argument but ignore it with IPv6, seems weird)

Reply:
Ok, makes sense. The `IsIPv6` method still feels unnecessary as opposed to checking `GetLocalIPAddress` for v6-ness. I see you now actually check the URN in that method, which I don't understand (and the method name doesn't seem fully applicable anymore).

Reply:
I agree that the name is questionable (maybe `IsIPv6GatewayDevice` or `ShouldUseIPv6Protocol` would be more accurate) but the reason that it now checks the URN instead of the local IP is that the `LocalIP` is only used for IPv4 now (another thing I should probably rename)

For IPv6, since it is common to have multiple IPv6 addresses (some of them temporary) on an interface so the `IGDService` now gets a reference to the interface and discovers it's IPv6 GUAs through that when AddPinhole is called. (We could rename that to `LocalIPv4`)